ipfs-cluster-service/main.go

```diff
@@ -251,17 +252,114 @@ configuration.
 					Name:  "upgrade",
 					Usage: "upgrade the IPFS Cluster state to the current version",
 					Description: `
-
 This command upgrades the internal state of the ipfs-cluster node 
 specified in the latest raft snapshot. The state format is migrated from the 
 version of the snapshot to the version supported by the current cluster version. 
 To successfully run an upgrade of an entire cluster, shut down each peer without
 removal, upgrade state using this command, and restart every peer.
 `,
 					Action: func(c *cli.Context) error {
-						err := upgrade()
+						err := locker.lock()
+						checkErr("acquiring execution lock", err)
+						defer locker.tryUnlock()
+
+						err = upgrade()
 						checkErr("upgrading state", err)
-						return err
+						return nil
+					},
+				},
+				{
+					Name:  "export",
+					Usage: "save the IPFS Cluster state to a json file",
+					Description: `
+This command reads the current cluster state and saves it as json for 
+human readability and editing.  Only state formats compatible with this
+version of ipfs-cluster-service can be exported.  By default this command
+prints the state to stdout.
+`,
+					Flags: []cli.Flag{
+						cli.StringFlag{
+							Name:  "file, f",
+							Value: "",
+							Usage: "sets an output file for exported state",
+						},
+					},
+					Action: func(c *cli.Context) error {
+						err := locker.lock()
+						checkErr("acquiring execution lock", err)
+						defer locker.tryUnlock()
+
+						err = export(c.String("file"))
+						checkErr("exporting state", err)
+						return nil
+					},
+				},
+				{
+					Name:  "import",
+					Usage: "load an IPFS Cluster state from an exported state file",
+					Description: `
+This command reads in an exported state file storing the state as a persistent
+snapshot to be loaded as the cluster state when the cluster peer is restarted.
+If an argument is provided, cluster will treat it as the path of the file to
+import.  If no argument is provided cluster will read json from stdin
+`,
+					Action: func(c *cli.Context) error {
+						err := locker.lock()
+						checkErr("acquiring execution lock", err)
+						defer locker.tryUnlock()
+
+						if !c.GlobalBool("force") {
+							logger.Info("Importing new state. All cluster peers must import the same state to avoid raft errors")
+							if !yesNoPrompt("The peer's state will be replaced.  Existing pins may be lost.  Continue? [y/n]") {
+								return nil
+							}
+						}
+
+						// Get the importing file path
+						importFile := c.Args().First()
+						var r io.ReadCloser
+						if importFile == "" {
+							r = os.Stdin
```

Comment:
If I'm correct, this should just look like the command is hanging when called without piping anything. Maybe it's worth doing a log.Info `Reading from stdin, Ctrl-D to finish` message there.

I don't know if there is an easy way to detect that stdin is interactively connected and only print that message in that case and not when something is piped, otherwise probably it won't hurt to print it in any case.